### PR TITLE
fix: Inject coder ssh configuration between comments

### PR DESCRIPTION
**Pull request by BrunoQuaresma**

### CODE REVIEW

**Comment by BrunoQuaresma** [comment]
Closes https://github.com/coder/vscode-coder/issues/42

File: src/remote.ts [hunk 1]
```diff
@@ -386,7 +388,27 @@ export class Remote {
...
     }
-    const parsedConfig = SSHConfig.parse(sshConfigRaw)
+
+    // We are going to extract only the coder config
+    let coderConfigRaw = ""
+    const startOfCoderConfig = sshConfigRaw.indexOf(startBlockComment)
+    const endOfCoderConfig = sshConfigRaw.lastIndexOf(endBlockComment)
+
+    // There is an existent coder config
+    if (startOfCoderConfig > -1 && endOfCoderConfig > -1) {
```
Comment (mafredri) [inline src/remote.ts]: We should also check `startOfCoderConfig < endOfCoderConfig`. See https://github.com/coder/coder/pull/5859.

File: src/remote.ts [hunk 2]
```diff
@@ -386,7 +388,27 @@ export class Remote {
       // Probably just doesn't exist!
       sshConfigRaw = ""
     }
-    const parsedConfig = SSHConfig.parse(sshConfigRaw)
+
+    // We are going to extract only the coder config
+    let coderConfigRaw = ""
+    const startOfCoderConfig = sshConfigRaw.indexOf(startBlockComment)
+    const endOfCoderConfig = sshConfigRaw.lastIndexOf(endBlockComment)
```
Comment (mafredri) [inline src/remote.ts]: This (lastIndexOf) seems a bit unsafe considering there could be malformed comments. We should ensure there's only one of each exists or ask the user to correct it.

File: src/remote.ts (hunk 1 above)
Comment (BrunoQuaresma) [reply src/remote.ts]: Good catch!

File: src/remote.ts (hunk 2 above)
Comment (BrunoQuaresma) [reply src/remote.ts]: A good one too!

**Comment by mafredri** [comment]
> Does it look right?

Yup, that sounds perfect @BrunoQuaresma. 👍🏻

**Comment by BrunoQuaresma** [comment]
I wanted to add some tests to the `SSHConfig` but I figure out the testing runner for this project needs some extra work so I decided to not do it on this PR.

File: src/SSHConfig.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    return `  ${text}`
+  }
+
+  private async save() {
+    await ensureDir(path.dirname(this.filePath))
+    return fs.writeFile(this.filePath, this.getRaw(), "utf-8")
+  }
+
+  private getRaw() {
+    if (!this.raw) {
```
Comment (mafredri) [inline src/SSHConfig.ts]: Should this check explicitly for undefined? Will match empty string as well.

File: src/SSHConfig.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    if (startBlockIndex === -1) {
+      throw new SSHConfigBadFormat("Start block not found")
+    }
+
+    if (startBlockIndex === -1) {
+      throw new SSHConfigBadFormat("End block not found")
+    }
+
+    if (endBlockIndex < startBlockIndex) {
+      throw new SSHConfigBadFormat("End block is starting before the start block")
```
Comment (mafredri) [inline src/SSHConfig.ts]:
```suggestion
      throw new SSHConfigBadFormat("Malformed config, end block is before start block")
```

File: src/SSHConfig.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  private appendBlock({ Host, ...otherValues }: SSHValues) {
+    const lines = [this.startBlockComment, `Host ${Host}`]
+    const keys = Object.keys(otherValues) as Array<keyof typeof otherValues>
+    keys.forEach((key) => {
+      lines.push(this.withIndentation(`${key} ${otherValues[key]}`))
+    })
+    lines.push(this.endBlockComment)
+    this.raw = `${this.raw}\n${lines.join("\n")}`
```
Comment (mafredri) [inline src/SSHConfig.ts]: Have you tested how this affects a config when a block is repeatedly removed and added back, or repeatedly updated/replaced?

For instance, when doing the above the newline count in the config should not constantly be growing. I.e. one risk is there are 100s of newlines at the end after a bunch of operations.

File: src/SSHConfig.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  private withIndentation(text: string) {
+    return `  ${text}`
```
Comment (mafredri) [inline src/SSHConfig.ts]: Suggestion (future improvement), parse config to figure out indentation used by other hosts (tabs, 2 spaces, 4 spaces etc), and use the same if unified across the file.

File: src/SSHConfig.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  private async save() {
+    await ensureDir(path.dirname(this.filePath))
```
Comment (mafredri) [inline src/SSHConfig.ts]: If ssh dir doesn't exist and we create it here, we should ensure it has good permissions. Usually `chmod 0700`, i.e. only owner has rwx permission, not group or everyone.

Likewise for `~/.ssh/config`, if we create it then a good permission would be `chmod 0600` (owner rw).

**Review by mafredri — COMMENTED** [review]
Looking good. Still had some pretty minor but probably important (for good user experience) nits.

File: src/SSHConfig.ts (hunk 3 above)
Comment (BrunoQuaresma) [reply src/SSHConfig.ts]: Good catch! And I didn't know JS would consider an empty string as false. 🤦

File: src/SSHConfig.ts (hunk 5 above)
Comment (BrunoQuaresma) [reply src/SSHConfig.ts]: Good catch!

**Review by mafredri — APPROVED** [review]
Thanks for fixing the nits. Code looks good to me, but I haven't tested the extension so might be good to have another set of eyes on it.

File: src/SSHConfig.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
+import { ensureDir } from "fs-extra"
```
Comment (kylecarbs) [inline src/SSHConfig.ts]: What do you think about camel casing this name instead? So there's uniformity with other files.

**Review by kylecarbs — APPROVED** [review]

File: src/SSHConfig.ts (hunk 8 above)
Comment (BrunoQuaresma) [reply src/SSHConfig.ts]: I just started it with capital case because it is a class, but I can do it.

**Pull request merged**